/*
  Copyright 2023 Google LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/

syntax = "proto3";
package qualtran;
import "qualtran/protos/args.proto";

message QBit {
}

message QAny {
    optional IntOrSympy bitsize = 1;
}

message QInt {
    optional IntOrSympy bitsize = 1;
}

message QIntOnesComp {
    optional IntOrSympy bitsize = 1;
}

message QUInt {
    optional IntOrSympy bitsize = 1;
}

message BoundedQUInt {
    optional IntOrSympy bitsize = 1;
    optional IntOrSympy iteration_length = 2;
}

message QFxp {
    optional IntOrSympy bitsize = 1;
    optional IntOrSympy num_frac = 2;
    optional bool signed = 3;
}

message QDataType {
  oneof val {
    QBit qbit = 1;
    QAny qany = 2;
    QInt qint = 3;
    QIntOnesComp qint_ones_comp = 4;
    QUInt quint = 5;
    BoundedQUInt bounded_quint = 6;
    QFxp qfxp = 7;
  }
}